File: mapbox_maps_flutter_mobile/ios/mapbox_maps_flutter/Sources/mapbox_maps_flutter/Classes/PerformanceStatisticsController.swift
import Foundation
@_spi(Experimental) import MapboxMaps

final class PerformanceStatisticsController: _PerformanceStatisticsApi {
    private let mapboxMap: MapboxMap
    private var collectionCancelable: AnyCancelable?
    private var performanceStatisticsListener: PerformanceStatisticsListener?

    init(mapboxMap: MapboxMap, messenger: SuffixBinaryMessenger) {
        self.mapboxMap = mapboxMap
        performanceStatisticsListener = PerformanceStatisticsListener(
            binaryMessenger: messenger.messenger,
            messageChannelSuffix: messenger.suffix
        )
    }

    func startPerformanceStatisticsCollection(options: PerformanceStatisticsOptions) throws {
        let options = options.toPerformanceStatisticsOptions()
        collectionCancelable = mapboxMap.collectPerformanceStatistics(options) { [weak self] statistics in
            self?.performanceStatisticsListener?.onPerformanceStatisticsCollected(
                statistics: statistics.toFLTPerformanceStatistics(),
                completion: { _ in }
            )
        }
    }

    func stopPerformanceStatisticsCollection() throws {
        collectionCancelable = nil
    }
}

File: mapbox_maps_flutter_mobile/android/src/main/kotlin/com/mapbox/maps/mapbox_maps/PerformanceStatisticsController.kt
package com.mapbox.maps.mapbox_maps

import com.mapbox.maps.MapboxMap
import com.mapbox.maps.mapbox_maps.pigeons.PerformanceStatisticsOptions
import com.mapbox.maps.mapbox_maps.pigeons._PerformanceStatisticsApi
import com.mapbox.maps.mapbox_maps.pigeons._PerformanceStatisticsListenerApi
import io.flutter.plugin.common.BinaryMessenger

class PerformanceStatisticsController(
  private val mapboxMap: MapboxMap,
  messenger: BinaryMessenger,
  channelSuffix: String
) : _PerformanceStatisticsApi {
  private val performanceStatisticsListener: _PerformanceStatisticsListenerApi =
    _PerformanceStatisticsListenerApi(messenger, channelSuffix)

  override fun startPerformanceStatisticsCollection(options: PerformanceStatisticsOptions) {
    // stop previous collection to match the behavior with iOS
    mapboxMap.stopPerformanceStatisticsCollection()

    mapboxMap.startPerformanceStatisticsCollection(options.toPerformanceStatisticsOptions()) { statistics ->
      performanceStatisticsListener.onPerformanceStatisticsCollected(statistics.toPerformanceStatistics()) { }
    }
  }

  override fun stopPerformanceStatisticsCollection() {
    mapboxMap.stopPerformanceStatisticsCollection()
  }
}